sklearn/feature_selection/mutual_info.py

```diff
@@ -0,0 +1,375 @@
+        elif 0 < self.n_features_to_select < 1:
+            return max(1, int(self.n_features_to_select * self._n_features))
+        else:
+            raise ValueError("`n_features_to_select` must be positive.")
+
+    def fit(self, X, y):
+        """Find the best features according to mutual information criterion.
+
+        Parameters
+        ----------
+        X : array_like or sparse matrix, shape (n_samples, n_features)
+            Feature matrix.
+        y : array_like, shape (n_samples,)
+            Target vector.
+
+        Returns
+        -------
+        self
+        """
+        X, y = check_X_y(X, y, accept_sparse='csc',
+                         y_numeric=not self.categorical_target)
+
+        n_features = X.shape[1]
+        self._n_features = n_features
+        n_features_to_select = self._get_n_features_to_select()
+
+        if isinstance(self.categorical_features, bool):
+            categorical_features = np.empty(n_features, dtype=bool)
+            categorical_features.fill(self.categorical_features)
+        else:
+            categorical_features = np.asarray(self.categorical_features)
+
+        if (not np.all(categorical_features) and
+                (self.use_redundancy or not self.categorical_target)):
+            X = scale(X, with_mean=False, copy=self.copy)
+
+        if not self.categorical_target:
+            y_std = np.std(y)
+            if y_std > 0:
+                y = y / y_std
+
+        if issparse(X):
+            iterate_columns = _iterate_csc_columns
+        else:
+            iterate_columns = _iterate_dense_columns
+
+        relevance = np.empty(n_features)
+        for i, xi in enumerate(iterate_columns(X)):
+            relevance[i] = _compute_mi(xi, y, categorical_features[i],
+                                       self.categorical_target,
+                                       self.n_neighbors)
+
+        if not self.use_redundancy:
+            self.ranks_ = np.argsort(-relevance)
+            return self
+
+        ranks = np.empty(n_features, dtype=int)
+        ranks.fill(n_features - 1)
+
+        redundancy = np.empty((n_features, n_features))
+
+        support = np.zeros(n_features, dtype=bool)
+        last_selected = np.argmax(relevance)
+        support[last_selected] = True
+        ranks[last_selected] = 0
+
+        for r in range(1, n_features_to_select):
+            selected = np.nonzero(support)[0]
+            candidates = np.nonzero(~support)[0]
+
+            x = _get_column(X, last_selected)
+            for i, y in zip(candidates, iterate_columns(X, candidates)):
+                mi = _compute_mi(x, y, categorical_features[last_selected],
+                                 categorical_features[i],
+                                 n_neighbors=self.n_neighbors)
+                redundancy[i, last_selected] = mi
+
+            scores_subset = (
+                relevance[candidates] -
+                np.mean(redundancy[np.ix_(candidates, selected)], axis=1))
```

Comment:
Seems like `np.ix_` is still necessary, i.e. `redundancy[np.ix_(~support, support)]`. Overall I think it's fine as it is.